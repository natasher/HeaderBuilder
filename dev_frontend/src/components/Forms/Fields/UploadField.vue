<template>
  <div class = "row">

    <div class = "col col-left">
      <span class = "title">{{ fieldName }}</span>
    </div>

    <div class = "col col-right">

      <div class = "mfn-field mfnf-upload">

        <input
          type   = "text"
          :value = "storedImageUrl"
          @input = "value => storedImageUrl = value.target.value" />

        <button
          v-if  = "storedImageUrl === ''"
          class = "mfn-button-generic add"
          @click.prevent = "openWPMedia">
            Add Image
        </button>
        <button
          v-else
          class = "mfn-button-generic remove"
          @click.prevent = "removeImg">
            Remove
        </button>

        <img
          :src = "storedImageUrl"
          v-if = "(/\.(gif|jpg|jpeg|tiff|png)$/i).test(storedImageUrl)" />

      </div>

    </div>

  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex'

export default {

  name: 'UploadField',

  props: {
    /**
     * The name of the field.
     * Displayed on the left column.
     */
    fieldName: {
      type: String,
      required: true,
    },
    /**
     * Name of the field used by WordPress internals.
     */
    wpId: {
      type: String,
      require: true,
    },
    /**
     * Prop determinates type of Margin field,
     * possible values: **""**, **row**, **grid**.
     */
    as: {
      type: String,
      required: false,
      default: '',
    },
    /**
     * When `as` prop is set to row, `row` indicates which one.
     * Then margin values will be set to given row.
     */
    row: {
      type: String,
      required: false,
      default: '',
    },
  },

  methods: {
    ...mapActions( 'fields', [ 'setModalFieldValue' ] ),

    /**
     * openWPMedia opens WordPress native media frame.
     */
    openWPMedia: function() {
      const that = this

      const upload_frame = (wp.media.frames.mfn_hb_gallery = wp.media({
        title  : 'Choose a File',
        library: {
          type: 'image',
        },
        button: {
          text: 'Select File',
        },
      }))

      upload_frame.on( 'select', function() {
        const attachment = upload_frame
          .state()
          .get('selection')
          .first()
          .toJSON()

        that.setModalFieldValue({
          name : that.wpId,
          value: attachment.url,
          as   : that.as  || '',
          row  : that.row || '',
        })
      })

      upload_frame.open()
    },

    removeImg: function() {
      this.setModalFieldValue({
        name : this.wpId,
        value: '',
        as   : this.as  || '',
        row  : this.row || '',
      })
    },
  },

  computed: {
    ...mapGetters( 'fields', [ 'getUploadFieldValue' ]),

    storedImageUrl: {
      get() {
        const img = this.getUploadFieldValue( this.wpId, this.as, this.row )

        return typeof img === 'undefined' ? '' : img
      },
      set(value) {
        this.setModalFieldValue({
          name : this.wpId,
          value: value,
          as   : this.as  || '',
          row  : this.row || '',
        })
      },
    },

    // TODO: usunąć?!
    // imageSelected: function() {
    //   return !jQuery.isEmptyObject( this.storedImageUrl )
    // },
  },
}
</script>
