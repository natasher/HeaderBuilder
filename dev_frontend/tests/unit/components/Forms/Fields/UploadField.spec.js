import { shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex          from 'vuex'
import flushPromises from 'flush-promises'
import merge         from 'lodash.merge'
import UploadField   from '../../../../../src/components/Forms/Fields/UploadField.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

const imgUrl = 'https://lorem.test/shitty.png'

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      fields: {
        namespaced: true,
        actions: {
          setModalFieldValue: jest.fn()
        },
        getters: {
          getUploadFieldValue: jest.fn().mockReturnValue(
            jest.fn().mockReturnValue( '' )
          ),
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    propsData: {
      fieldName: 'Test Field',
      wpId     : 'testField',
    },
  }

  return shallowMount( UploadField, merge( defaultMountingOptions, overrides ))
}

describe('TextField.vue', () => {

  describe('snapshots:', () => {

    test('should match snapshot if field does NOT contain image url', async () => {
      const wrapper = createWrapper()
      await flushPromises()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match snapshot if field contain image url', () => {
      const innerGetter         = jest.fn().mockReturnValue( imgUrl )
      const getUploadFieldValue = jest.fn().mockReturnValue( innerGetter )
      const store = createStore({
        modules: {
          fields: {
            getters: {
              getUploadFieldValue,
            }
          }
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  // test('should display `Add Image` button and hide `Remove` button if no image URL provided')
  // test('should display `Remove` button and hide `Add Image` button if image URL is provided')
  // test('should allow user to type image URL - external image scenario')
  // test('should open Wordpress media when the User click `Add Image`')
  // test('should remove image URL if the User clicks `Remove` button')
  // test('should display <img /> if image URL end with gif, jpg, jpeg, tiff, png')

})