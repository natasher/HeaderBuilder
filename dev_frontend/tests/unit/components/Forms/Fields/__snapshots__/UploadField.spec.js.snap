// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField.vue snapshots: should match snapshot if field contain image url 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-upload"
    >
      <input
        type="text"
      />
       
      <button
        class="mfn-button-generic remove"
      >
        
          Remove
      
      </button>
       
      <img
        src="https://lorem.test/shitty.png"
      />
    </div>
  </div>
</div>
`;

exports[`TextField.vue snapshots: should match snapshot if field does NOT contain image url 1`] = `
<div
  class="row"
>
  <div
    class="col col-left"
  >
    <span
      class="title"
    >
      Test Field
    </span>
  </div>
   
  <div
    class="col col-right"
  >
    <div
      class="mfn-field mfnf-upload"
    >
      <input
        type="text"
      />
       
      <button
        class="mfn-button-generic add"
      >
        
          Add Image
      
      </button>
       
      <!---->
    </div>
  </div>
</div>
`;
